### PR TITLE
Feature/vsdspub 59 following ldes client

### PR DESCRIPTION
**Pull request by Yalz**

### CODE REVIEW

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java [hunk 1]
```diff
@@ -1,42 +1,66 @@
...
-        this.nextPageToProcess = initialPageToProcess;
     }
 
-    public boolean processItem(String item) {
-        return processedItems.add(item);
+    public boolean processMember(String member) {
+        return processedItems.add(member);
     }
 
+    public boolean hasPagesToProcess() {
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:36]: It's much more comfortable to rename this to hasNext()

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/LdesClient.java [hunk 2]
```diff
@@ -44,37 +46,57 @@ public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
...
     }
 
     @Override
     public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
         flowManager = new FlowManager(session);
-
-        JsonObject ldesInput = ldesInputRestClient.retrieveLdesInput(stateManager.getNextPageToProcess());
-        if(ldesInput == null) {
-            return;
+        if (stateManager.hasPagesToProcess()) {
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/LdesClient.java:54]: stateManager.hasPagesToProcess already checks if there are more pages to process.

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/LdesClient.java [hunk 3]
```diff
@@ -44,37 +46,57 @@ public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
...
+
+        // Next Pages
+        model.listStatements(null, model.createProperty("https://w3id.org/tree#", "node"), (String) null)
+                .forEach(statement -> stateManager.addNewPageToProcess(statement.getObject().toString()));
+
+        if (!stateManager.hasPagesToProcess()) {
+            stateManager.setFullyReplayed(true);
+        }
     }
 }
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/LdesClient.java:102]: newline

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java [hunk 4]
```diff
@@ -1,42 +1,66 @@
...
+        this.pagesToProcess.add(initialPageToProcess);
         this.processedItems = new HashSet<>();
+        this.processedPages = new HashSet<>();
         this.treeDirection = TreeDirection.valueOf(treeDirection);
-        this.nextPageToProcess = initialPageToProcess;
     }
 
-    public boolean processItem(String item) {
-        return processedItems.add(item);
+    public boolean processMember(String member) {
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:32]: I agree with the rename to processMember, but why not also rename processedItems to processedMembers ?

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java (hunk 1 above)
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:36]: Much more comfortable is to follow the Iterable interface convention: hasNext()

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java [hunk 5]
```diff
@@ -1,42 +1,66 @@
...
+        return processedItems.add(member);
     }
 
+    public boolean hasPagesToProcess() {
+        return !pagesToProcess.isEmpty();
+    }
     public String getNextPageToProcess() {
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:39]: Iterable: next()

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java [hunk 6]
```diff
@@ -1,42 +1,66 @@
...
-import be.vlaanderen.vsds.ldesclient.nifi.processors.models.LdesPage;
 import be.vlaanderen.vsds.ldesclient.nifi.processors.models.TreeDirection;
 
-import java.util.*;
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.Queue;
+import java.util.Set;
 
 public class StateManager {
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:10]: Iterator and Iterable should be implemented for this.

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java [hunk 7]
```diff
@@ -1,42 +1,66 @@
...
+    public boolean processMember(String member) {
+        return processedItems.add(member);
     }
 
+    public boolean hasPagesToProcess() {
+        return !pagesToProcess.isEmpty();
+    }
     public String getNextPageToProcess() {
-        return nextPageToProcess;
+        if(!hasPagesToProcess()) {
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:40]: No need for this. If the queue is empty, then poll() will return null and you can catch it there.
Throwing a RuntimeException is wrong: if the queue is empty, we've done our work.

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java [hunk 8]
```diff
@@ -1,42 +1,66 @@
...
+            throw new RuntimeException("No more pages to process");
+        }
+
+        String pageUrl = pagesToProcess.poll();
+        processedPages.add(pageUrl);
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:45]: if pageUrl != null ...

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java [hunk 9]
```diff
@@ -1,42 +1,66 @@
...
-    private String nextPageToProcess;
-    private final Set<String> processedItems;
-    private final TreeDirection treeDirection;
+    private final Queue<String> pagesToProcess;
+    protected final Set<String> processedPages;
+    protected final Set<String> processedItems;
+    private TreeDirection treeDirection;
     private boolean fullyReplayed;
 
+    public StateManager(String initialPageToProcess) {
```
Comment (sverholen) [inline nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:17]: Page -> Fragment

**Review by sverholen — APPROVED** [review]
Good job on getting this working !
Will merge to ensure the team can start using the code, but needs changes.

General remark:
- We need to discuss setting up a model module (e.g. for LdesMember, LdesFragment).
- We need to discuss using a consistent vocabulary (page is a hydra word, fragment is LDES)

File: nifi-processors/pom.xml [hunk 10]
```diff
@@ -35,25 +35,29 @@
             <version>${nifi-utils.version}</version>
         </dependency>
         <dependency>
-            <groupId>org.apache.httpcomponents</groupId>
-            <artifactId>httpclient</artifactId>
-            <version>${apache-httpclient.version}</version>
+            <groupId>org.projectlombok</groupId>
```
[the remark this answers is not in the document]
Comment (Yalz) [reply nifi-processors/pom.xml:38]: I think I eventually don't use it anymore. Yet I know Wouter is a big fan of it. So maybe let's bring it up in a team meeting
As part of my cleanup i'll delete this dependency for now

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Yalz) [reply nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:32]: very true point. adjusting!

File: nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java (hunk 1 above)
Comment (Yalz) [reply nifi-processors/src/main/java/be/vlaanderen/vsds/ldesclient/nifi/processors/services/StateManager.java:36]: To use hasNext on the pagesToProcess Queue, I can agreee with. 
However this state manager keeps more than just the pages. So I'd hold on to the method name though

File: nifi-processors/src/main/resources/META-INF/services/org.apache.nifi.processor.Processor [hunk 11]
```diff
@@ -12,5 +12,4 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-be.vlaanderen.vsds.ldesclient.nifi.processors.ReplicateLdesStream
 be.vlaanderen.vsds.ldesclient.nifi.processors.LdesClient
```
[the remark this answers is not in the document]
Comment (Yalz) [reply nifi-processors/src/main/resources/META-INF/services/org.apache.nifi.processor.Processor:15]: Not sure what is meant here?

**Pull request merged**